### PR TITLE
Gradient checkpointing experiments

### PR DESCRIPTION
**Pull request by NeoLegends**
I'm using this PR to run the CI. Albert don't bother reviewing yet.

### CODE REVIEW

File: returnn/frontend/_backend.py [hunk 1]
```diff
@@ -1293,6 +1293,22 @@ def tensor_array_get_item(cls, tensor_array: TensorArrayType, index: Union[int,
             return tensor_array[index]
         raise NotImplementedError
 
+    @classmethod
+    def gradient_checkpoint(cls, fn: Callable[[Tensor], Tensor], *args: List[Tensor]) -> Tensor:
```
Comment (albertz) [inline returnn/frontend/_backend.py:1297]: Is this exactly like `recompute_grad`? I think i would rather then call it `recompute_grad`.

File: returnn/frontend/linear.py [hunk 2]
```diff
@@ -40,6 +41,43 @@ def __call__(self, source: Tensor) -> Tensor:
         return out
 
 
+class LinearWithWeightDropout(rf.Module):
```
Comment (albertz) [inline returnn/frontend/linear.py:44]: Why do you add this? We don't want to have specialized variants for every single existing module for every single variant of reparameterization (e.g. weight dropout, variational noise, weight normalization, ...). See #1518.

File: returnn/frontend/_backend.py (hunk 1 above)
Comment (albertz) [inline returnn/frontend/_backend.py:1297]: I think the typing is wrong in multiple ways. If you accept multiple args for `fn`, shouldn't it be `Callable[[Tensor, ...], Tensor]` or so? And also, I think the type for `args` is just `Tensor` (I think you specify the type of individual elements).

File: tests/test_rf_base.py [hunk 3]
```diff
@@ -231,6 +231,32 @@ def _forward_step(*, model: _Net, extern_data: TensorDict):
...
+
+        def __call__(self, x: Tensor) -> Tensor:
+            return self.linear(x)
+
+    # noinspection PyShadowingNames
+    def _forward_step(*, model: _Net, extern_data: TensorDict):
+        out = model(extern_data["data"])
+        out.mark_as_default_output()
+
+    run_model(extern_data, lambda *, epoch, step: _Net(), _forward_step)
```
Comment (albertz) [inline tests/test_rf_base.py:257]: Note that currently, by default, we test the TF-layers backend and the PyTorch backend. But you only have implemented the direct TF backend. I'm not saying that you also must implement the TF-layers backend, just that you maybe want to disable the TF-layers backend here, and maybe instead enable the direct TF backend.

Comment (albertz) [inline tests/test_rf_base.py:257]: Also note, this test will actually not test at all the weight dropout. It will operate in forward mode, i.e. disable any dropout. But even if you enable training, what you also would want to check is whether it did the grad checkpointing actually correctly. Although I have no idea how to test this. In TF, I have similar tests by really investigating the computation graph. Not sure how you can investigate the gradient tape in PT. Or how you can verify in any way that the gradient checkpointing worked.

File: tests/test_rf_base.py [hunk 4]
```diff
@@ -231,6 +231,32 @@ def _forward_step(*, model: _Net, extern_data: TensorDict):
     run_model(extern_data, lambda *, epoch, step: _Net(), _forward_step)
 
 
+def test_linear():
```
Comment (albertz) [inline tests/test_rf_base.py:234]: The name of the test is misleading. This is not about testing `linear`. I guess you want to test weight dropout and/or gradient checkpointing here.

File: returnn/frontend/ckpt.py [hunk 5]
```diff
@@ -0,0 +1,31 @@
+"""
```
Comment (albertz) [inline returnn/frontend/ckpt.py:1]: The file name is misleading. `ckpt` is usually for (model) checkpoints.

In TF, there is a file `custom_gradient.py`, where you find functions like `custom_gradient` and `recompute_grad`. We actually already have `returnn/frontend/gradient.py` where you also find some similar related functions. I think the best fit would be there.

File: returnn/frontend/ckpt.py [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+from __future__ import annotations
+from typing import Callable, TypeVar
+
+__all__ = ["gradient_checkpoint"]
+
+
+T = TypeVar("T")
+
+
+def gradient_checkpoint(fn: Callable[..., T], *args: ...) -> T:
```
Comment (albertz) [inline returnn/frontend/ckpt.py:14]: I think `recompute_grad` is a better name.

File: tests/test_rf_base.py (hunk 4 above)
Comment (albertz) [inline tests/test_rf_base.py:234]: If you only want to test gradient checkpointing, `test_rf_gradient.py` is a better place for the test.

**Comment by albertz** [comment]
> I'm using this PR to run the CI.

It's fine that you put intermediate work into a draft PR. But you don't need to use the CI to test things. You can simply run the relevant tests locally. I usually run the relevant tests inside a debugger (e.g. PyCharm) so I can directly see potential issues. It's actually a nice way to debug and develop, to write the test case and then run it in the debugger.

**Comment by albertz** [comment]
Hey, what's the state here? Did you test this? Does it work? I.e. it doesn't store the intermediate activations in memory?

**Comment by NeoLegends** [comment]
Hey, no new results yet.

**Comment by albertz** [comment]
> Hey, no new results yet.

But what was the state? What were the old results? Did it work? Does it not store the intermediate activations in memory anymore? Or you don't know?

**Comment by albertz** [comment]
Also, I don't exactly understand how this API here could be used for variational noise and/or weight dropout. Can you give an example?

**Comment by NeoLegends** [comment]
Superseeded by #1559
